### ISSUE TITLE
Abnormal join sql can not be cancelled

### ISSUE THREAD

**Issue by ljcui** [ClickHouse/ClickHouse#28635]
(you don't have to strictly follow this form)

**Describe the unexpected behaviour**

sql like this：
`
select w1.id
FROM
(
    SELECT 'test' as id FROM numbers(10000000)
) w1
global left join 
(
    SELECT 'test' as id FROM numbers(10000000)
) w2 
on w1.id = w2.id
limit 10;
`
It runs for several hours and consumes a lot of CPU，must be restarted to recover.

**Comment by ljcui** [comment]
The configuration item `max_rows_in_join` does not take effect

**Comment by alexey-milovidov** [comment]
This operation will produce 100 000 000 000 records.

**Comment by alexey-milovidov** [comment]
@vdimir Let's check some limits in `joinRightColumns`?

Similar checks as between blocks.
We can do it periodically if it's ALL JOIN and the number of records per block is large.

**Comment by ljcui** [comment]
To fix this issue: 
    Add a new limit configuration item to allow the `for` loop in `joinRightColumns` to exit
Is this okay?
I am trying to modify the code like this ...

**Comment by vdimir** [comment]
> To fix this issue:
> Add a new limit configuration item to allow the `for` loop in `joinRightColumns` to exit.
> Is this okay?

Yes, I suppose we can check size limits in `joinRightColumns`, like in `addJoinedBlock`
https://github.com/ClickHouse/ClickHouse/blob/7a4a0b0edef0ad6e0aa662cd3b90c3f4acf796e7/src/Interpreters/HashJoin.cpp#L712

I think we shouldn't introduce new option for this, maybe reuse existing one in some way, or find reasonable heuristic.

> I am trying to modify the code like this ...

Great, let me know if you'll need some help
